compiler/src/iree/compiler/Dialect/Flow/Transforms/DequantizationMatmulFuse.cpp

```diff
@@ -0,0 +1,190 @@
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/Utils/Utils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+//-----------------------------------------------------------//
+//                        Utility
+//-----------------------------------------------------------//
+
+// Creates a new flow.dipatch.region op and places the
+// passed ops inside as long as the dequant op is a
+// producer for the matmul op
+static LogicalResult fuseDequantAndMatmul(RewriterBase &rewriter,
+                                          Operation *dequant,
+                                          Operation *matmul) {
+  
+  Flow::DispatchRegionOp regionOp =
+      matmul->getParentOfType<DispatchRegionOp>();
+  if (!regionOp) {
+    FailureOr<DispatchRegionOp> maybeRegionOp =
+        wrapOpInDispatchRegion(rewriter, matmul);
+    if (failed(maybeRegionOp))
+      return failure();
+    regionOp = maybeRegionOp.value();
+  }
+
+  FailureOr<DispatchRegionOp> maybeFusedRegionOp =
+      movePrecedingOpsIntoDispatchRegion(rewriter, dequant, regionOp);
+  if (failed(maybeFusedRegionOp))
+    return failure();
+
+  return success();
+}
+
+// Checks if the passed op is a contraction on grouped input
+// with 2 reduction dimensions
+static LogicalResult isGroupedContractionOp(linalg::GenericOp genericOp) {
+  unsigned numLoops = genericOp.getNumLoops();
+  linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(genericOp.getOperation());
+  if (numLoops == 0)
+    return failure();
+  if (!linalg::isaContractionOpInterface(linalgOp))
+    return failure();
+  if (genericOp.getNumReductionLoops() != 2)
+    return failure();
+  return success();
+}
+
+// Checks if the passed op is a dequantization on grouped input
+// This expects the dequantization op to have a body like:
+//    arith.extui
+//    arith.uitofp
+//    arith.subf
+//    arith.mulf
+static LogicalResult isGroupedDequantizationOp(linalg::GenericOp genericOp) {
+  // Check for 1 result, and 2 (input, scales) or 3 (input, scales, zero points) inputs
+  if (genericOp.getNumDpsInits() != 1)
+    return failure();
+  if (genericOp.getNumDpsInputs() != 2 && genericOp.getNumDpsInputs() != 3)
+    return failure();
+
+  // Check that the rank is at least 3 and all loops are parallel
+  unsigned numLoops = genericOp.getNumLoops();
+  unsigned numParallelLoops = genericOp.getNumParallelLoops();
+  if (numLoops < 3)
+    return failure();
+  if (numLoops != numParallelLoops)
+    return failure();
+
+  auto body = genericOp.getBlock();
+  SmallVector<Operation *, 16> bodyOps;
```

Comment:
Instead of this, you really want to match a DAG of the operations in the body. You can use `matchPatterns` and `m_Op`, etc. like here https://github.com/llvm/llvm-project/blob/66bb6b4fdaedf1c4f5bc19a6fc7d2f5c33aa9471/mlir/lib/Dialect/Arith/IR/ArithOps.cpp#L791 to match the DAG of operations.


The Structured op matchers can maybe make this more powerful, but I am also still learning this, so we can do that as a follow up.